controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,373 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	l "github.com/redhat-appstudio/build-service/pkg/logs"
+	"golang.org/x/exp/slices"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"knative.dev/pkg/apis"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strings"
+
+	applicationapi "github.com/redhat-appstudio/application-api/api/v1alpha1"
+	tektonapi "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+)
+
+const (
+	contextTimeout = 300 * time.Second
+	// PipelineRunTypeLabelName contains the type of the PipelineRunTypeLabelName.
+	PipelineRunTypeLabelName = "pipelines.appstudio.openshift.io/type"
+	// PipelineRunBuildType is the type denoting a build PipelineRun.
+	PipelineRunBuildType = "build"
+	// PacEventTypeLabelName represents the current event type
+	PacEventTypeLabelName = "pipelinesascode.tekton.dev/event-type"
+	PacEventPushType      = "push"
+	ImageUrlParamName     = "IMAGE_URL"
+	ImageDigestParamName  = "IMAGE_DIGEST"
+
+	// PipelineRunComponentLabelName is the label denoting the application.
+	PipelineRunComponentLabelName = "appstudio.openshift.io/component"
+
+	NudgeProcessedAnnotationName = "build.appstudio.openshift.io/component-nudge-processed"
+	NudgeFinalizer               = "build.appstudio.openshift.io/build-nudge-finalizer"
+	FailureCountAnnotationName   = "build.appstudio.openshift.io/build-nudge-failures"
+
+	// AlreadyNudgedAnnotationName this is populated in the case of an error when some components were successful
+	AlreadyNudgedAnnotationName = "build.appstudio.openshift.io/nudged-components"
+
+	ComponentNudgedEventType      = "ComponentNudged"
+	ComponentNudgeFailedEventType = "ComponentNudgeFailed"
+	MaxAttempts                   = 3
+	FailureRetryTime              = time.Minute * 5 // We retry after 5 minutes on failure
+)
+
+// ComponentDependencyUpdateReconciler reconciles a PipelineRun object
+type ComponentDependencyUpdateReconciler struct {
+	client.Client
+	Scheme         *runtime.Scheme
+	EventRecorder  record.EventRecorder
+	UpdateFunction UpdateComponentDependenciesFunction
+}
```

Comment:
What's the consideration for adding a function pointer `UpdateFunction`? Easy for writing tests?

Reply:
To allow it to be easily mocked in tests.